src/app/clusters/group-key-mgmt-server/group-key-mgmt-server.cpp

```diff
@@ -0,0 +1,410 @@
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <app/CommandHandler.h>
+#include <app/common/gen/att-storage.h>
+#include <app/common/gen/attribute-id.h>
+#include <app/common/gen/attribute-type.h>
+#include <app/common/gen/attributes/Accessors.h>
+#include <app/common/gen/cluster-id.h>
+#include <app/common/gen/command-id.h>
+#include <app/util/af.h>
+#include <map>
+#include <set>
+#include <string.h>
+#include <tuple>
+#include <utility>
+
+constexpr size_t GROUP_KEY_ROOT_SIZE  = 16;
+constexpr size_t GROUP_KEY_EPOCH_SIZE = 8;
+constexpr uint8_t GROUP_KEY_SET_SIZE  = 3;
+
+using namespace chip;
+
+struct RootKey
+{
+    bool active                             = false;
+    uint8_t value[GROUP_KEY_ROOT_SIZE]      = { 0 };
+    uint8_t startTime[GROUP_KEY_EPOCH_SIZE] = { 0 };
+    uint8_t policy                          = 0;
+};
+
+struct KeySet
+{
+    uint8_t current = 0;
+    RootKey epoch[GROUP_KEY_SET_SIZE];
+};
+
+static std::map<std::tuple<uint16_t, uint16_t>, KeySet *> sGroupKeys;
+static std::set<std::tuple<uint16_t, GroupId, uint16_t>> sGroupStates;
+
+//------------------------------------------------------------------------------
+// Local functions
+//------------------------------------------------------------------------------
+
+static EmberAfStatus revokeKeys(uint16_t fabricIndex, uint16_t keySetIndex);
+static EmberAfStatus revokeKeys(uint16_t fabricIndex);
+
+static uint8_t nextEpoch(uint8_t epoch)
+{
+    if (epoch < GROUP_KEY_SET_SIZE - 1)
+    {
+        return epoch + static_cast<uint8_t>(1);
+    }
+    else
+    {
+        return 1;
+    }
+}
+
+static uint8_t previousEpoch(uint8_t epoch)
+{
+    if (epoch > 0)
+    {
+        return epoch - static_cast<uint8_t>(1);
+    }
+    else
+    {
+        return GROUP_KEY_SET_SIZE - 1;
+    }
+}
+
+static EmberAfStatus printKeys()
+{
+    emberAfGroupKeyManagementClusterPrint("GROUP KEY SETS(%u)", sGroupKeys.size());
+
+    std::map<std::tuple<uint16_t, uint16_t>, KeySet *>::iterator it;
+    uint16_t fabricIndex = 0;
+    uint16_t keySet      = 0;
+
+    for (it = sGroupKeys.begin(); it != sGroupKeys.end(); ++it)
+    {
+        fabricIndex = std::get<0>(it->first);
+        keySet      = std::get<1>(it->first);
+
+        emberAfGroupKeyManagementClusterPrint("  fabric:%u, set:%u", fabricIndex, keySet);
+        uint8_t index     = it->second->current < GROUP_KEY_SET_SIZE ? it->second->current : 0;
+        RootKey & current = it->second->epoch[index];
```

Comment:
Consider a helper variable for `it->second` that gives it a proper name, such as `keySet`